// ng2-bs3-modal/fesm5/ng2-bs3-modal.js
import { Injectable, Component, Input, Output, EventEmitter, ElementRef, HostBinding, NgZone, Directive, Optional, NgModule } from '@angular/core';
import { fromEvent, Observable, Subject, of, merge, zip } from 'rxjs';
import { filter, map, share, tap, take } from 'rxjs/operators';
import { __spread } from 'tslib';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @enum {number} */
var BsModalHideType = {
    Close: 0,
    Dismiss: 1,
    Backdrop: 2,
    Keyboard: 3,
    RouteChange: 4,
    Destroy: 5,
};
BsModalHideType[BsModalHideType.Close] = 'Close';
BsModalHideType[BsModalHideType.Dismiss] = 'Dismiss';
BsModalHideType[BsModalHideType.Backdrop] = 'Backdrop';
BsModalHideType[BsModalHideType.Keyboard] = 'Keyboard';
BsModalHideType[BsModalHideType.RouteChange] = 'RouteChange';
BsModalHideType[BsModalHideType.Destroy] = 'Destroy';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var BsModalSize = /** @class */ (function () {
    function BsModalSize() {
    }
    /**
     * @param {?} size
     * @return {?}
     */
    BsModalSize.isValidSize = /**
     * @param {?} size
     * @return {?}
     */
    function (size) {
        return size && (size === BsModalSize.Small || size === BsModalSize.Large);
    };
    BsModalSize.Small = 'sm';
    BsModalSize.Large = 'lg';
    return BsModalSize;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
var EVENT_SUFFIX = 'ng2-bs3-modal';
/** @type {?} */
var KEYUP_EVENT_NAME = "keyup." + EVENT_SUFFIX;
/** @type {?} */
var CLICK_EVENT_NAME = "click." + EVENT_SUFFIX;
/** @type {?} */
var SHOW_EVENT_NAME = "show.bs.modal." + EVENT_SUFFIX;
var BsModalService = /** @class */ (function () {
    function BsModalService() {
        var _this = this;
        this.modals = [];
        this.$body = jQuery(document.body);
        this.onBackdropClose$ = fromEvent(this.$body, CLICK_EVENT_NAME).pipe(filter(function (e) { return jQuery(e.target).is('.modal'); }), map(function () { return BsModalHideType.Backdrop; }), share());
        this.onKeyboardClose$ = fromEvent(this.$body, KEYUP_EVENT_NAME).pipe(filter(function (e) { return e.which === 27; }), map(function () { return BsModalHideType.Keyboard; }), share());
        this.onModalStack$ = fromEvent(this.$body, SHOW_EVENT_NAME).pipe(tap(function () {
            /** @type {?} */
            var zIndex = 1040 + (10 * jQuery('.modal:visible').length);
            jQuery(_this).css('z-index', zIndex);
            setTimeout(function () {
                jQuery('.modal-backdrop').not('.modal-stack').css('z-index', zIndex - 1).addClass('modal-stack');
            }, 0);
        }), share());
    }
    /**
     * @param {?} modal
     * @return {?}
     */
    BsModalService.prototype.add = /**
     * @param {?} modal
     * @return {?}
     */
    function (modal) {
        this.modals.push(modal);
    };
    /**
     * @param {?} modal
     * @return {?}
     */
    BsModalService.prototype.remove = /**
     * @param {?} modal
     * @return {?}
     */
    function (modal) {
        /** @type {?} */
        var index = this.modals.indexOf(modal);
        if (index > -1) {
            this.modals.splice(index, 1);
        }
    };
    /**
     * @return {?}
     */
    BsModalService.prototype.focusNext = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var visible = this.modals.filter(function (m) { return m.visible; });
        if (visible.length) {
            this.$body.addClass('modal-open');
            visible[visible.length - 1].focus();
        }
    };
    /**
     * @return {?}
     */
    BsModalService.prototype.dismissAll = /**
     * @return {?}
     */
    function () {
        return Promise.all(this.modals.map(function (m) {
            return m.dismiss();
        }));
    };
    BsModalService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    BsModalService.ctorParameters = function () { return []; };
    return BsModalService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
var EVENT_SUFFIX$1 = 'ng2-bs3-modal';
/** @type {?} */
var SHOW_EVENT_NAME$1 = "show.bs.modal." + EVENT_SUFFIX$1;
/** @type {?} */
var SHOWN_EVENT_NAME = "shown.bs.modal." + EVENT_SUFFIX$1;
/** @type {?} */
var HIDE_EVENT_NAME = "hide.bs.modal." + EVENT_SUFFIX$1;
/** @type {?} */
var HIDDEN_EVENT_NAME = "hidden.bs.modal." + EVENT_SUFFIX$1;
/** @type {?} */
var LOADED_EVENT_NAME = "loaded.bs.modal." + EVENT_SUFFIX$1;
/** @type {?} */
var DATA_KEY = 'bs.modal';
var BsModalComponent = /** @class */ (function () {
    function BsModalComponent(element, service, zone) {
        var _this = this;
        this.element = element;
        this.service = service;
        this.zone = zone;
        this.overrideSize = null;
        this.onInternalClose$ = new Subject();
        this.subscriptions = [];
        this.visible = false;
        this.showing = false;
        this.hiding = false;
        this.animation = true;
        this.backdrop = true;
        this.keyboard = true;
        this.onShow = new EventEmitter();
        this.onOpen = new EventEmitter();
        this.onHide = new EventEmitter();
        this.onClose = new EventEmitter();
        this.onDismiss = new EventEmitter();
        this.onLoaded = new EventEmitter();
        this.setVisible = function (isVisible) {
            return function () {
                _this.visible = isVisible;
                _this.showing = false;
                _this.hiding = false;
            };
        };
        this.setOptions = function (options) {
            /** @type {?} */
            var backdrop = options.backdrop;
            if (typeof backdrop === 'string' && backdrop !== 'static') {
                backdrop = true;
            }
            if (options.backdrop !== undefined) {
                _this.options.backdrop = backdrop;
            }
            if (options.keyboard !== undefined) {
                _this.options.keyboard = options.keyboard;
            }
        };
        this.service.add(this);
        this.init();
    }
    Object.defineProperty(BsModalComponent.prototype, "options", {
        get: /**
         * @return {?}
         */
        function () {
            if (!this.$modal) {
                this.init();
            }
            return this.$modal.data(DATA_KEY).options;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BsModalComponent.prototype, "fadeClass", {
        get: /**
         * @return {?}
         */
        function () { return this.animation; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BsModalComponent.prototype, "modalClass", {
        get: /**
         * @return {?}
         */
        function () { return true; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BsModalComponent.prototype, "roleAttr", {
        get: /**
         * @return {?}
         */
        function () { return 'dialog'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BsModalComponent.prototype, "tabindexAttr", {
        get: /**
         * @return {?}
         */
        function () { return '-1'; },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    BsModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.wireUpEventEmitters();
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.$dialog = this.$modal.find('.modal-dialog');
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.setOptions({
            backdrop: this.backdrop,
            keyboard: this.keyboard
        });
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.onInternalClose$.next(BsModalHideType.Destroy);
        return this.destroy();
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.focus = /**
     * @return {?}
     */
    function () {
        this.$modal.trigger('focus');
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.routerCanDeactivate = /**
     * @return {?}
     */
    function () {
        this.onInternalClose$.next(BsModalHideType.RouteChange);
        return this.destroy();
    };
    /**
     * @param {?=} size
     * @return {?}
     */
    BsModalComponent.prototype.open = /**
     * @param {?=} size
     * @return {?}
     */
    function (size) {
        this.overrideSize = null;
        if (BsModalSize.isValidSize(size)) {
            this.overrideSize = size;
        }
        return this.show().toPromise();
    };
    /**
     * @param {?=} value
     * @return {?}
     */
    BsModalComponent.prototype.close = /**
     * @param {?=} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        this.onInternalClose$.next(BsModalHideType.Close);
        return this.hide().pipe(tap(function () { return _this.onClose.emit(value); })).toPromise().then(function () { return value; });
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.dismiss = /**
     * @return {?}
     */
    function () {
        this.onInternalClose$.next(BsModalHideType.Dismiss);
        return this.hide().toPromise();
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.getCssClasses = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var classes = [];
        if (this.isSmall()) {
            classes.push('modal-sm');
        }
        if (this.isLarge()) {
            classes.push('modal-lg');
        }
        if (this.cssClass) {
            classes.push(this.cssClass);
        }
        return classes.join(' ');
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.isSmall = /**
     * @return {?}
     */
    function () {
        return this.overrideSize !== BsModalSize.Large
            && this.size === BsModalSize.Small
            || this.overrideSize === BsModalSize.Small;
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.isLarge = /**
     * @return {?}
     */
    function () {
        return this.overrideSize !== BsModalSize.Small
            && this.size === BsModalSize.Large
            || this.overrideSize === BsModalSize.Large;
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.show = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.visible && !this.hiding) {
            return of(null);
        }
        this.showing = true;
        return Observable.create(function (o) {
            _this.onShown$.pipe(take(1)).subscribe(function (next) {
                o.next(next);
                o.complete();
            });
            _this.transitionFix();
            _this.$modal.modal('show');
        });
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.transitionFix = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // Fix for shown.bs.modal not firing when .fade is present
        // https://github.com/twbs/bootstrap/issues/11793
        if (this.animation) {
            setTimeout(function () {
                _this.$modal.trigger('focus').trigger(SHOWN_EVENT_NAME);
            }, jQuery.fn.modal['Constructor'].TRANSITION_DURATION);
        }
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.hide = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.visible && !this.showing) {
            return of(null);
        }
        this.hiding = true;
        return Observable.create(function (o) {
            _this.onHidden$.pipe(take(1)).subscribe(function (next) {
                o.next(next);
                o.complete();
            });
            _this.$modal.modal('hide');
        });
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.init = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.$modal = jQuery(this.element.nativeElement);
        this.$modal.appendTo(document.body);
        this.$modal.modal({
            show: false
        });
        this.onShowEvent$ = fromEvent(this.$modal, SHOW_EVENT_NAME$1);
        this.onShownEvent$ = fromEvent(this.$modal, SHOWN_EVENT_NAME);
        this.onHideEvent$ = fromEvent(this.$modal, HIDE_EVENT_NAME);
        this.onHiddenEvent$ = fromEvent(this.$modal, HIDDEN_EVENT_NAME);
        this.onLoadedEvent$ = fromEvent(this.$modal, LOADED_EVENT_NAME);
        /** @type {?} */
        var onClose$ = merge(this.onInternalClose$, this.service.onBackdropClose$, this.service.onKeyboardClose$);
        this.onHide$ = zip(this.onHideEvent$, onClose$).pipe(map(function (x) { return ({ event: x[0], type: x[1] }); }));
        this.onHidden$ = zip(this.onHiddenEvent$, onClose$).pipe(map(function (x) { return x[1]; }), tap(this.setVisible(false)), tap(function () { return _this.service.focusNext(); }), share());
        this.onShown$ = this.onShownEvent$.pipe(tap(this.setVisible(true)), share());
        this.onDismiss$ = this.onHidden$.pipe(filter(function (x) { return x !== BsModalHideType.Close; }));
        // Start watching for events
        // Start watching for events
        (_a = this.subscriptions).push.apply(_a, __spread([
            this.onShown$.subscribe(function () { }),
            this.onHidden$.subscribe(function () { }),
            this.service.onModalStack$.subscribe(function () { })
        ]));
        var _a;
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.wireUpEventEmitters = /**
     * @return {?}
     */
    function () {
        this.wireUpEventEmitter(this.onShow, this.onShowEvent$);
        this.wireUpEventEmitter(this.onOpen, this.onShown$);
        this.wireUpEventEmitter(this.onHide, this.onHide$);
        this.wireUpEventEmitter(this.onDismiss, this.onDismiss$);
        this.wireUpEventEmitter(this.onLoaded, this.onLoadedEvent$);
    };
    /**
     * @template T
     * @param {?} emitter
     * @param {?} stream$
     * @return {?}
     */
    BsModalComponent.prototype.wireUpEventEmitter = /**
     * @template T
     * @param {?} emitter
     * @param {?} stream$
     * @return {?}
     */
    function (emitter, stream$) {
        var _this = this;
        if (emitter.observers.length === 0) {
            return;
        }
        /** @type {?} */
        var sub = stream$.subscribe(function (next) {
            _this.zone.run(function () {
                emitter.next(next);
            });
        });
        this.subscriptions.push(sub);
    };
    /**
     * @return {?}
     */
    BsModalComponent.prototype.destroy = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return this.hide().pipe(tap(function () {
            _this.service.remove(_this);
            _this.subscriptions.forEach(function (s) { return s.unsubscribe(); });
            _this.subscriptions = [];
            if (_this.$modal) {
                _this.$modal.data(DATA_KEY, null);
                _this.$modal.remove();
                _this.$modal = null;
            }
        })).toPromise();
    };
    BsModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'bs-modal',
                    template: "\n        <div class=\"modal-dialog\" [ngClass]=\"getCssClasses()\">\n            <div class=\"modal-content\">\n                <ng-content></ng-content>\n            </div>\n        </div>\n    "
                },] },
    ];
    /** @nocollapse */
    BsModalComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: BsModalService },
        { type: NgZone }
    ]; };
    BsModalComponent.propDecorators = {
        animation: [{ type: Input }],
        backdrop: [{ type: Input }],
        keyboard: [{ type: Input }],
        size: [{ type: Input }],
        cssClass: [{ type: Input }],
        onShow: [{ type: Output }],
        onOpen: [{ type: Output }],
        onHide: [{ type: Output }],
        onClose: [{ type: Output }],
        onDismiss: [{ type: Output }],
        onLoaded: [{ type: Output }],
        fadeClass: [{ type: HostBinding, args: ['class.fade',] }],
        modalClass: [{ type: HostBinding, args: ['class.modal',] }],
        roleAttr: [{ type: HostBinding, args: ['attr.role',] }],
        tabindexAttr: [{ type: HostBinding, args: ['attr.tabindex',] }]
    };
    return BsModalComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var BsModalHeaderComponent = /** @class */ (function () {
    function BsModalHeaderComponent(modal) {
        this.modal = modal;
        this.showDismiss = false;
    }
    BsModalHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'bs-modal-header',
                    template: "\n        <div class=\"modal-header\">\n            <button *ngIf=\"showDismiss\" type=\"button\" class=\"close\" aria-label=\"Dismiss\" (click)=\"modal.dismiss()\">\n                <span aria-hidden=\"true\">&times;</span>\n            </button>\n            <ng-content></ng-content>\n        </div>\n    "
                },] },
    ];
    /** @nocollapse */
    BsModalHeaderComponent.ctorParameters = function () { return [
        { type: BsModalComponent }
    ]; };
    BsModalHeaderComponent.propDecorators = {
        showDismiss: [{ type: Input }]
    };
    return BsModalHeaderComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var BsModalBodyComponent = /** @class */ (function () {
    function BsModalBodyComponent() {
    }
    BsModalBodyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'bs-modal-body',
                    template: "\n        <div class=\"modal-body\">\n            <ng-content></ng-content>\n        </div>\n    "
                },] },
    ];
    return BsModalBodyComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var BsModalFooterComponent = /** @class */ (function () {
    function BsModalFooterComponent(modal) {
        this.modal = modal;
        this.showDefaultButtons = false;
        this.dismissButtonLabel = 'Dismiss';
        this.closeButtonLabel = 'Close';
    }
    BsModalFooterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'bs-modal-footer',
                    template: "\n        <div class=\"modal-footer\">\n            <ng-content></ng-content>\n            <button *ngIf=\"showDefaultButtons\" type=\"button\" class=\"btn btn-default\" (click)=\"modal.dismiss()\">\n                {{dismissButtonLabel}}\n            </button>\n            <button *ngIf=\"showDefaultButtons\" type=\"button\" class=\"btn btn-primary\" (click)=\"modal.close()\">\n                {{closeButtonLabel}}\n              </button>\n        </div>\n    "
                },] },
    ];
    /** @nocollapse */
    BsModalFooterComponent.ctorParameters = function () { return [
        { type: BsModalComponent }
    ]; };
    BsModalFooterComponent.propDecorators = {
        showDefaultButtons: [{ type: Input }],
        dismissButtonLabel: [{ type: Input }],
        closeButtonLabel: [{ type: Input }]
    };
    return BsModalFooterComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var BsAutofocusDirective = /** @class */ (function () {
    function BsAutofocusDirective(el, modal) {
        var _this = this;
        this.el = el;
        this.modal = modal;
        if (modal) {
            this.modal.onOpen.subscribe(function () {
                _this.el.nativeElement.focus();
            });
        }
    }
    BsAutofocusDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[autofocus]'
                },] },
    ];
    /** @nocollapse */
    BsAutofocusDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: BsModalComponent, decorators: [{ type: Optional }] }
    ]; };
    return BsAutofocusDirective;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var BsModalModule = /** @class */ (function () {
    function BsModalModule() {
    }
    BsModalModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        BsModalComponent,
                        BsModalHeaderComponent,
                        BsModalBodyComponent,
                        BsModalFooterComponent,
                        BsAutofocusDirective
                    ],
                    providers: [
                        BsModalService
                    ],
                    exports: [
                        BsModalComponent,
                        BsModalHeaderComponent,
                        BsModalBodyComponent,
                        BsModalFooterComponent,
                        BsAutofocusDirective
                    ]
                },] },
    ];
    return BsModalModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { BsModalService, BsModalComponent, BsModalHeaderComponent, BsModalBodyComponent, BsModalFooterComponent, BsModalHideType, BsModalSize, BsModalModule, BsAutofocusDirective as ɵa };
