// ng2-bs3-modal/esm2015/modal/footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { BsModalComponent } from './modal.component';
export class BsModalFooterComponent {
    /**
     * @param {?} modal
     */
    constructor(modal) {
        this.modal = modal;
        this.showDefaultButtons = false;
        this.dismissButtonLabel = 'Dismiss';
        this.closeButtonLabel = 'Close';
    }
}
BsModalFooterComponent.decorators = [
    { type: Component, args: [{
                selector: 'bs-modal-footer',
                template: `
        <div class="modal-footer">
            <ng-content></ng-content>
            <button *ngIf="showDefaultButtons" type="button" class="btn btn-default" (click)="modal.dismiss()">
                {{dismissButtonLabel}}
            </button>
            <button *ngIf="showDefaultButtons" type="button" class="btn btn-primary" (click)="modal.close()">
                {{closeButtonLabel}}
              </button>
        </div>
    `
            },] },
];
/** @nocollapse */
BsModalFooterComponent.ctorParameters = () => [
    { type: BsModalComponent }
];
BsModalFooterComponent.propDecorators = {
    showDefaultButtons: [{ type: Input }],
    dismissButtonLabel: [{ type: Input }],
    closeButtonLabel: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    BsModalFooterComponent.prototype.showDefaultButtons;
    /** @type {?} */
    BsModalFooterComponent.prototype.dismissButtonLabel;
    /** @type {?} */
    BsModalFooterComponent.prototype.closeButtonLabel;
    /** @type {?} */
    BsModalFooterComponent.prototype.modal;
}
